File: sdks/csharp/WebhookVerification.cs
using System;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;

namespace HookRelay
{
    /// <summary>
    /// Result of webhook verification.
    /// </summary>
    public class VerificationResult
    {
        public bool Valid { get; set; }
        public object? Payload { get; set; }
        public string? Error { get; set; }

        public static VerificationResult Invalid(string error) =>
            new VerificationResult { Valid = false, Error = error };

        public static VerificationResult ValidResult(object payload) =>
            new VerificationResult { Valid = true, Payload = payload };
    }

    /// <summary>
    /// Webhook signature verification for HookRelay.
    ///
    /// Supports both simple HMAC-SHA256 verification and Standard Webheaders
    /// (Svix-compatible) verification with timestamp tolerance.
    /// Also supports Svix headers (svix-id, svix-signature, svix-timestamp) as fallback.
    /// </summary>
    public static class WebhookVerification
    {
        private const int DefaultToleranceSecs = 300;

        /// <summary>
        /// Verify a webhook from HTTP headers with automatic header detection.
        /// Supports both Standard Webhooks and Svix headers.
        /// </summary>
        public static VerificationResult VerifyWebhookFromHeaders(
            string payload, System.Net.WebHeaderCollection headers, string secret, int toleranceSecs = DefaultToleranceSecs)
        {
            var msgId = headers["webhook-id"];
            var timestamp = headers["webhook-timestamp"];
            var signatureHeader = headers["webhook-signature"];

            if (string.IsNullOrEmpty(msgId) || string.IsNullOrEmpty(timestamp) || string.IsNullOrEmpty(signatureHeader))
            {
                msgId = msgId ?? headers["svix-id"];
                timestamp = timestamp ?? headers["svix-timestamp"];
                signatureHeader = signatureHeader ?? headers["svix-signature"];
            }

            return VerifyWebhook(payload, msgId, timestamp, signatureHeader, secret, toleranceSecs);
        }

        /// <summary>
        /// Verify a webhook signature using HMAC-SHA256.
        /// </summary>
        /// <param name="payload">The raw request body</param>
        /// <param name="signature">The signature from the X-Hookrelay-Signature header</param>
        /// <param name="secret">The endpoint's signing secret (starts with "whsec_")</param>
        /// <returns>true if the signature is valid</returns>
        public static bool VerifySignature(string payload, string signature, string secret)
        {
            if (string.IsNullOrEmpty(payload) || string.IsNullOrEmpty(signature) || string.IsNullOrEmpty(secret))
                return false;

            var expectedHex = signature.StartsWith("sha256=") ? signature[7..] : signature;

            try
            {
                using var hmac = new HMACSHA256(Encoding.UTF8.GetBytes(secret));
                var hash = hmac.ComputeHash(Encoding.UTF8.GetBytes(payload));
                var computed = Convert.ToHexString(hash).ToLowerInvariant();

                return CryptographicOperations.FixedTimeEquals(
                    Encoding.UTF8.GetBytes(computed),
                    Encoding.UTF8.GetBytes(expectedHex));
            }
            catch
            {
                return false;
            }
        }

        /// <summary>
        /// Verify a webhook using Standard Webheaders headers (Svix-compatible).
        /// </summary>
        public static VerificationResult VerifyWebhook(
            string payload, string? msgId, string? timestamp,
            string? signatureHeader, string secret, int toleranceSecs = DefaultToleranceSecs)
        {
            if (string.IsNullOrEmpty(msgId))
                return VerificationResult.Invalid("Missing webhook-id header");
            if (string.IsNullOrEmpty(timestamp))
                return VerificationResult.Invalid("Missing webhook-timestamp header");
            if (string.IsNullOrEmpty(signatureHeader))
                return VerificationResult.Invalid("Missing webhook-signature header");
            if (string.IsNullOrEmpty(payload))
                return VerificationResult.Invalid("Missing request body");

            // Validate timestamp
            if (!long.TryParse(timestamp, out var ts))
                return VerificationResult.Invalid("Invalid webhook timestamp");

            var now = DateTimeOffset.UtcNow.ToUnixTimeSeconds();
            var age = Math.Abs(now - ts);

            if (age > toleranceSecs)
                return VerificationResult.Invalid(
                    $"Webhook timestamp expired: {age}s old (tolerance: {toleranceSecs}s)");

            // Compute expected signature
            var signedContent = $"{msgId}.{timestamp}.{payload}";
            var secretBytes = DecodeSecret(secret);

            try
            {
                using var hmac = new HMACSHA256(secretBytes);
                var hash = hmac.ComputeHash(Encoding.UTF8.GetBytes(signedContent));
                var expectedSig = Convert.ToBase64String(hash);
                var expectedFull = $"v1,{expectedSig}";

                // Check each signature in the header (space-separated)
                var signatures = signatureHeader.Split(' ');
                var verified = false;

                foreach (var sig in signatures)
                {
                    var trimmed = sig.Trim();
                    if (!trimmed.StartsWith("v1,")) continue;

                    if (trimmed.Length == expectedFull.Length &&
                        CryptographicOperations.FixedTimeEquals(
                            Encoding.UTF8.GetBytes(trimmed),
                            Encoding.UTF8.GetBytes(expectedFull)))
                    {
                        verified = true;
                        break;
                    }
                }

                if (!verified)
                    return VerificationResult.Invalid("Invalid webhook signature");
            }
            catch
            {
                return VerificationResult.Invalid("Signature computation failed");
            }

            // Parse payload
            try
            {
                var parsed = JsonSerializer.Deserialize<JsonElement>(payload);
                return VerificationResult.ValidResult(parsed);
            }
            catch
            {
                return VerificationResult.ValidResult(payload);
            }
        }

        private static byte[] DecodeSecret(string secret)
        {
            var stripped = secret.StartsWith("whsec_") ? secret[6..] : secret;
            try
            {
                return Convert.FromBase64String(stripped);
            }
            catch
            {
                return Encoding.UTF8.GetBytes(secret);
            }
        }
    }
}

File: sdks/php/src/WebhookVerification.php
<?php

declare(strict_types=1);

namespace HookRelay;

/**
 * Webhook signature verification for HookRelay.
 *
 * Supports both simple HMAC-SHA256 verification and Standard Webheaders
 * (Svix-compatible) verification with timestamp tolerance.
 */
class WebhookVerification
{
    private const DEFAULT_TOLERANCE_SECS = 300;

    /**
     * Verify a webhook signature using HMAC-SHA256.
     *
     * @param string $payload   The raw request body
     * @param string $signature The signature from the X-Hookrelay-Signature header
     * @param string $secret    The endpoint's signing secret (starts with "whsec_")
     * @return bool true if the signature is valid
     */
    public static function verifySignature(string $payload, string $signature, string $secret): bool
    {
        if (empty($payload) || empty($signature) || empty($secret)) {
            return false;
        }

        $expectedHex = str_starts_with($signature, 'sha256=')
            ? substr($signature, 7)
            : $signature;

        $computed = hash_hmac('sha256', $payload, $secret);

        return hash_equals($computed, $expectedHex);
    }

    /**
     * Verify a webhook using Standard Webheaders headers (Svix-compatible).
     *
     * @param string $payload         The raw request body
     * @param string|null $msgId      The webhook-id header
     * @param string|null $timestamp  The webhook-timestamp header
     * @param string|null $signatureHeader The webhook-signature header
     * @param string $secret          The endpoint's signing secret
     * @param int    $toleranceSecs   Max age in seconds (default: 300)
     * @return array{valid: bool, payload?: mixed, error?: string}
     */
    public static function verifyWebhook(
        string $payload,
        ?string $msgId,
        ?string $timestamp,
        ?string $signatureHeader,
        string $secret,
        int $toleranceSecs = self::DEFAULT_TOLERANCE_SECS
    ): array {
        if (empty($msgId)) {
            return ['valid' => false, 'error' => 'Missing webhook-id header'];
        }
        if (empty($timestamp)) {
            return ['valid' => false, 'error' => 'Missing webhook-timestamp header'];
        }
        if (empty($signatureHeader)) {
            return ['valid' => false, 'error' => 'Missing webhook-signature header'];
        }
        if (empty($payload)) {
            return ['valid' => false, 'error' => 'Missing request body'];
        }

        // Validate timestamp
        $ts = (int) $timestamp;
        if ($ts === 0) {
            return ['valid' => false, 'error' => 'Invalid webhook timestamp'];
        }

        $now = time();
        $age = abs($now - $ts);

        if ($age > $toleranceSecs) {
            return [
                'valid' => false,
                'error' => sprintf('Webhook timestamp expired: %ds old (tolerance: %ds)', $age, $toleranceSecs),
            ];
        }

        // Compute expected signature
        $signedContent = "{$msgId}.{$timestamp}.{$payload}";
        $secretBytes = self::decodeSecret($secret);

        $expectedSig = base64_encode(
            hash_hmac('sha256', $signedContent, $secretBytes, true)
        );
        $expectedFull = "v1,{$expectedSig}";

        // Check each signature in the header (space-separated)
        $signatures = explode(' ', $signatureHeader);
        $verified = false;

        foreach ($signatures as $sig) {
            $trimmed = trim($sig);
            if (!str_starts_with($trimmed, 'v1,')) continue;

            if (hash_equals($trimmed, $expectedFull)) {
                $verified = true;
                break;
            }
        }

        if (!$verified) {
            return ['valid' => false, 'error' => 'Invalid webhook signature'];
        }

        // Parse payload
        $parsed = json_decode($payload, true);
        if ($parsed !== null) {
            return ['valid' => true, 'payload' => $parsed];
        }

        return ['valid' => true, 'payload' => $payload];
    }

    /**
     * Decode a Standard Webhooks secret.
     */
    private static function decodeSecret(string $secret): string
    {
        $stripped = str_starts_with($secret, 'whsec_')
            ? substr($secret, 6)
            : $secret;

        $decoded = base64_decode($stripped, true);
        return $decoded !== false ? $decoded : $secret;
    }
}
